smart-contract-verifier/smart-contract-verifier-server/src/types/batch_verification.rs

```diff
@@ -0,0 +1,156 @@
+        inner_contracts.push(inner_contract);
+    }
+    Ok(inner_contracts)
+}
+
+pub fn from_proto_compiler_version_to_inner(proto: &str) -> Result<Version, Status> {
+    Version::from_str(proto)
+        .map_err(|err| Status::invalid_argument(format!("Invalid compiler version: {}", err)))
+}
+
+pub fn from_proto_solidity_multi_part_content_to_inner(
+    sources: BTreeMap<String, String>,
+    evm_version: Option<String>,
+    optimization_runs: Option<u32>,
+    libraries: BTreeMap<String, String>,
+) -> Result<solidity::multi_part::MultiFileContent, Status> {
+    let sources = sources
+        .into_iter()
+        .map(|(file, content)| (PathBuf::from(file), content))
+        .collect();
+
+    let evm_version = evm_version
+        .as_ref()
+        .map(|value| foundry_compilers::EvmVersion::from_str(value))
+        .transpose()
+        .map_err(|err| Status::invalid_argument(format!("Invalid evm version: {}", err)))?;
+
+    Ok(solidity::multi_part::MultiFileContent {
+        sources,
+        evm_version,
+        optimization_runs: optimization_runs.map(|value| value as usize),
+        contract_libraries: (!libraries.is_empty()).then_some(libraries),
+    })
+}
+
+pub fn compilation_error(message: impl Into<String>) -> Response<BatchVerifyResponse> {
+    Response::new(BatchVerifyResponse {
+        verification_result: Some(
+            proto::batch_verify_response::VerificationResult::CompilationFailure(
+                CompilationFailure {
+                    message: message.into(),
+                },
+            ),
+        ),
+    })
+}
+
+pub fn process_batch_error(error: BatchError) -> Result<Response<BatchVerifyResponse>, Status> {
+    match error {
+        BatchError::VersionNotFound(_) => Err(Status::invalid_argument(error.to_string())),
+        BatchError::Compilation(_) => {
+            let response = compilation_error(error.to_string());
+            Ok(response)
+        }
+        BatchError::Internal(_) => Err(Status::internal(error.to_string())),
+    }
+}
+
+pub fn process_verification_results(
+    results: Vec<BatchVerificationResult>,
+) -> Result<Response<BatchVerifyResponse>, Status> {
+    let items = results
+        .into_iter()
+        .map(process_verification_result)
+        .collect::<Result<Vec<_>, _>>()?;
+    let response = BatchVerifyResponse {
+        verification_result: Some(
+            proto::batch_verify_response::VerificationResult::ContractVerificationResults(
+                proto::batch_verify_response::ContractVerificationResults { items },
+            ),
+        ),
+    };
+    Ok(Response::new(response))
+}
+
+fn process_verification_result(
+    result: BatchVerificationResult,
+) -> Result<proto::ContractVerificationResult, Status> {
+    let verification_result = match result {
+        BatchVerificationResult::Failure(_) => {
+            proto::contract_verification_result::VerificationResult::Failure(
+                proto::ContractVerificationFailure {},
```

Comment:
so we did't give any reason why verification failed, why?

Reply:
There is no main reason why verification failed. Different contracts resultant from compilation have different reasons. So, if verification failed it basically means that no contract corresponded to the bytecodes we are trying to verify (neither creation nor runtime), and there is no more details we can provide.

I had an idea to return the particular reason for each contract compiled but for now it seemed too much, imo